### ISSUE TITLE
aws_emr_cluster: parameters passed to bootstrap_action are in the wrong order

### ISSUE THREAD

**Issue by entelente** [hashicorp/terraform#10691]
### Terraform Version
0.7.13

### Affected Resource(s)
- aws_emr_cluster

### Terraform Configuration Files
```hcl
bootstrap_action {
path = "s3://bucket/script.sh"
    name = "scrip_name"
    args = ["1", "2", "3", "4", "5", "6", "7", "8", "9"]
}
```

### Output
```hcl
    bootstrap_action.#:                                 "1"
    bootstrap_action.2281593937.args.#:                 "9"
    bootstrap_action.2281593937.args.1790921346:        "7"
    bootstrap_action.2281593937.args.1842515611:        "3"
    bootstrap_action.2281593937.args.2212294583:        "1"
    bootstrap_action.2281593937.args.2226203566:        "5"
    bootstrap_action.2281593937.args.2366072709:        "9"
    bootstrap_action.2281593937.args.4088798008:        "4"
    bootstrap_action.2281593937.args.4194326291:        "8"
    bootstrap_action.2281593937.args.450215437:         "2"
    bootstrap_action.2281593937.args.498629140:         "6"
    bootstrap_action.2281593937.name:                   "script_name"
```

### Expected Behavior
Parameters passed in the correct order to the script.

### Actual Behavior
Parameters are passed to the script in an order they appeared in the provided output.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Create an aws emr resource with a bootstrap script that accepts multiple parameters
2. Bootstrap script will list the parameters it gets
3. `terraform apply`

**Comment by catsby** [comment]
Hey @ddcprg – I'm not sure this is an issue, at least not as presented. The `output` above is how the parameters are stored, but not necessarily how they are presented to the EMR service. They're stored as a `Set`, so the key is the hash of the contents. 

Can you confirm for me they are in the correct order in the EMR Console? If you're finding that they are presented to the EMR service in the incorrect order then we should apply patch #11049 but I'm not yet seeing that to be the case, can you confirm?

**Comment by catsby** [comment]
I meant to mention @entelente as well there, either way, can someone show me a configuration that demonstrates incorrect ordering? Our acc test is creating clusters in the correct order, I'd just like to confirm. Thanks!

**Comment by harpermann** [comment]
We're having the ordering problem.  Here is the args line in TF:

 args = [ "--install-boto", "--install-bluetalon", "--debug", "--efs-dns%20${aws_efs_mount_target.gep_emr_efs_mount_az1.dns_name}", "--efs-mount%20/test" ]

Here is the startExec line from /var/log/bootstrap_actions/1/controller on the EMR master after it starts.

2017-01-10T23:00:31.220Z INFO startExec '/emr/instance-controller/lib/bootstrap-actions/1/gbootstrap_emr.sh --efs-mount%20/predix --install-boto --install-bluetalon --efs-dns%20us-east-1c.fs-fa6eb7b3.efs.us-east-1.amazonaws.com --debug'

Since the args are out of order, I have to put args with values in one set of quotes and use %20 instead of a space as the args get single quotes with a space.  Can't escape (\ ) the space as TF doesn't allow the escape char in args.

[jenkins@ip-10-83-2-135 analytics_vpc]$ tf version
Terraform v0.7.13

Thanks for looking at this!

**Comment by ddcprg** [comment]
Hi @catsby I've just hit this bug during an update of my resources (using Terraform 0.8.2) - there are quite a few parameters involved here so I'll try to put it as plain as possible:

```
resource "aws_s3_bucket_object" "healthcheck-script" {
  bucket     = "my-bucket"
  key        = "scripts/healthcheck_service.py"
  source     = "healthcheck.py"
}

resource "aws_s3_bucket_object" "healthcheck-installer" {
  bucket     = "my-bucket"
  key        = "scripts/init_healthcheck.sh"
  source     = "init_healthcheck.sh"
}

resource "aws_emr_cluster" "important-service" {
  depends_on           = ["aws_s3_bucket_object.healthcheck-script", "aws_s3_bucket_object.healthcheck-installer"]
  count                = 2
  name                 = "important-service-${count.index}"
  release_label        = "emr-5.0.0"
  applications         = ["Hadoop", "Hive", "HCatalog"]
  log_uri              = "s3://my-bucket/important-service/"
  master_instance_type = "m1.medium"
  core_instance_count  = 1
  service_role         = "EMR_DefaultRole"

  ec2_attributes {
    key_name                          = "my-ec2-key"
    subnet_id                         = "subnet-..."
    emr_managed_master_security_group = "sg-..."
    additional_master_security_groups = "sg-..."
    emr_managed_slave_security_group  = "sg-..."
    additional_slave_security_groups  = "sg-..."
    service_access_security_group     = "sg-..."
    instance_profile                  = "arn:aws:iam::${var.account-id}:instance-profile/EMR_EC2_DefaultRole"
  }

  bootstrap_action {
    path = "s3://elasticmapreduce/bootstrap-actions/run-if"
    name = "init_healthcheck"
    args = ["instance.isMaster=true",
            "s3://my-bucket/scripts/init_healthcheck.sh",
            "s3://my-bucket/scripts/"]
  }
}
```

Here, `s3://elasticmapreduce/bootstrap-actions/run-if` should download `s3://my-bucket/scripts/init_healthcheck.sh` and execute:

```
s3://my-bucket/scripts/init_healthcheck.sh s3://my-bucket/scripts/
```

But the cluster creation fails during bootstrap with this error:

`Terminated with errorsOn the master instance (i-01xxxxxxxxxxx), bootstrap action 1 `

And in the logs is clear the parameters are inverted - it tries to download the third parameters which is a "directory" in the bucket:

```
Attempting to download file 's3://my-bucket/scripts/'
Error downloading file from Amazon S3
```

I hope this helps to identify the issue above.

**Comment by entelente** [comment]
Hi @catsby!

Sorry for the late answer, I had to fabricate some code to reproduce the issue. The original code was modified to use named parameters as a workaround.

TF file:
```
provider "aws" {
  shared_credentials_file = "./creds"
  profile					        = "saml"
  region					        = "us-west-2"  
}

resource "aws_emr_cluster" "test" {
  count                = 1
  name                 = "jstrasszer-test"
  release_label        = "emr-5.0.0"
  applications         = ["Hadoop", "Hive"]
  log_uri              = "s3n://epam-jstrasszer/testlog/"
  master_instance_type = "m4.large"
  core_instance_type   = "m1.small"
  core_instance_count  = 1
  service_role         = "EMR_DefaultRole"

  ec2_attributes {
    key_name                          = "jstrasszer"
    subnet_id                         = "subnet-439f9421"
    emr_managed_master_security_group = "sg-e6eb6d80"
    additional_master_security_groups = "sg-12b78074"
    emr_managed_slave_security_group  = "sg-0aeb6d6c"
    additional_slave_security_groups  = "sg-12b78074"
    instance_profile                  = "arn:aws:iam::261830790520:instance-profile/EMR_EC2_DefaultRole"

  }
  
  bootstrap_action=[
    {
      path = "s3://epam-jstrasszer/test.sh"
      name = "test"
      args = ["1",
              "2",
              "3",
              "4",
              "5",
              "6",
              "7",
              "8",
              "9",
              "10" ]
    }
  ]
}
```

Contents of test.sh:
```
#!/bin/bash

echo $@
```

After apply this was the console output:
```
$ terraform apply
aws_emr_cluster.test: Creating...
  applications.#:                                     "" => "2"
  applications.2086606022:                            "" => "Hive"
  applications.3006569794:                            "" => "Hadoop"
  bootstrap_action.#:                                 "" => "1"
  bootstrap_action.3243030435.args.#:                 "" => "10"
  bootstrap_action.3243030435.args.1790921346:        "" => "7"
  bootstrap_action.3243030435.args.1842515611:        "" => "3"
  bootstrap_action.3243030435.args.2212294583:        "" => "1"
  bootstrap_action.3243030435.args.2226203566:        "" => "5"
  bootstrap_action.3243030435.args.2366072709:        "" => "9"
  bootstrap_action.3243030435.args.2707236321:        "" => "10"
  bootstrap_action.3243030435.args.4088798008:        "" => "4"
  bootstrap_action.3243030435.args.4194326291:        "" => "8"
  bootstrap_action.3243030435.args.450215437:         "" => "2"
  bootstrap_action.3243030435.args.498629140:         "" => "6"
  bootstrap_action.3243030435.name:                   "" => "test"
  bootstrap_action.3243030435.path:                   "" => "s3://epam-jstrasszer/test.sh"
  cluster_state:                                      "" => "<computed>"
  core_instance_count:                                "" => "1"
  core_instance_type:                                 "" => "m1.small"
  ec2_attributes.#:                                   "" => "1"
  ec2_attributes.0.additional_master_security_groups: "" => "sg-12b78074"
  ec2_attributes.0.additional_slave_security_groups:  "" => "sg-12b78074"
  ec2_attributes.0.emr_managed_master_security_group: "" => "sg-e6eb6d80"
  ec2_attributes.0.emr_managed_slave_security_group:  "" => "sg-0aeb6d6c"
  ec2_attributes.0.instance_profile:                  "" => "arn:aws:iam::261830790520:instance-profile/EMR_EC2_DefaultRole"
  ec2_attributes.0.key_name:                          "" => "jstrasszer"
  ec2_attributes.0.subnet_id:                         "" => "subnet-439f9421"
  log_uri:                                            "" => "s3n://epam-jstrasszer/testlog/"
  master_instance_type:                               "" => "m4.large"
  master_public_dns:                                  "" => "<computed>"
  name:                                               "" => "jstrasszer-test"
  release_label:                                      "" => "emr-5.0.0"
  service_role:                                       "" => "EMR_DefaultRole"
  visible_to_all_users:                               "" => "true"
```

And here is the output of test.sh from the logs:
```
7 3 1 5 9 10 4 8 2 6
```

**Comment by ddcprg** [comment]
Hi @catsby any updates on this ticket?

**Comment by ddcprg** [comment]
Hey guys, I'm sorry for chasing you up but I wonder whether this can be looked at.

**Comment by ddcprg** [comment]
Any updates on this issue?

**Comment by lattwood** [comment]
@catsby I can confirm this is occuring still, and they're in the wrong order in the AWS Console

**Comment by catsby** [comment]
I've opened a fix for this in #13580

**Comment by TimOrme** [comment]
Trying to piece together the history for this a bit, but it seems that this has already been released. 

Based on the release dates it seems that this went out in Apr 2017 and v0.11.7 went out in April 2018, implying that it should be addressed there? We're running v0.11.7 and are still encountering this issue however.  Any ideas?

Sorry if this is poor etiquette by the way, if its helpful to open a new bug, let me know.
